### PR TITLE
Complete rework of our CI

### PR DESCRIPTION
**Pull request by AnotherDaniel**
This is to address #51, and will be a multi-stage PR (a few individual commits with change sets building up the whole picture).

Step 1 - enable cargo-deny, adding deny.toml, and changing a build.rs dep for avoid strange licensed crates in our dep tree

Step 2 - remove Cargo.lock from our .gitignore and add a couple of template files to enable cargo-about

Step 3 - introduce the all-new github workflows

### CODE REVIEW

File: .github/workflows/x-build.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      NEXTEST_EXPERIMENTAL_LIBTEST_JSON: 1
+    strategy:
+      matrix:
+        include:
+          - name: linux-amd64
+            runner: ubuntu-latest
+            target: x86_64-unknown-linux-gnu
+          - name: linux-arm64
+            runner: ubuntu-latest
+            target: aarch64-unknown-linux-gnu
```
Comment (PLeVasseur) [inline .github/workflows/x-build.yaml:41]: :+1: 

Was about to ask for this yesterday on the issue, but you beat me to the punch :slightly_smiling_face:

**Comment by PLeVasseur** [comment]
Hey @AnotherDaniel -- opened an issue for adding cross-compiling: #73 

Do you think we should tackle that as a part of this PR or a new one after this one merges?

**Comment by AnotherDaniel** [comment]
God this workspace setup with locked-in required workflows is driving up the wall in this context here. Who has admin rights to this repo and can just fix this? Should be related to this feature: https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/

**Comment by PLeVasseur** [comment]
> Who has admin rights to this repo and can just fix this? Should be related to this feature: https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/

I dunno if commiters get admin rights. If so => @sophokles73, if not => @stevenhartley

**Comment by stevenhartley** [comment]
> > Who has admin rights to this repo and can just fix this? Should be related to this feature: https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/
> 
> I dunno if committers get admin rights. If so => @sophokles73, if not => @stevenhartley

Hi @AnotherDaniel , that is me and I think anyone else. We use branch protection rules that require certain workflows to be successful to run. These rules are turned on/off using otterdog in the https://github.com/eclipse-uprotocol/.eclipsefdn project. As you can see in https://github.com/eclipse-uprotocol/.eclipsefdn/blob/main/otterdog/eclipse-uprotocol.jsonnet#L355 we require three workflows to be successful so if you change the names of the workflows (or delete them), the branch protection rules will fail. We can temporarily turn them off, fix the workflows, then turn them back on if you like? It is a 10 min change for me, let me know if you want me to do that for your project. You can also fork this project, make the change, then create a PR to merge it back in.

**Comment by AnotherDaniel** [comment]
@stevenhartley then I think the path forward would be to turn off these rules, merge this PR and turn them on again?

Would work for me, I can't do any of these things but would pick things up once these workflows are active.

**Comment by stevenhartley** [comment]
Change is pending, once merge you only have to retrigger the change by pushing something (black space or something): https://github.com/eclipse-uprotocol/.eclipsefdn/pull/46

**Review by stevenhartley — APPROVED** [review]

**Comment by stevenhartley** [comment]
@AnotherDaniel it is all passing and looking great, please let me know if it is ready to merge or if you would like others to chime in as well.

**Comment by AnotherDaniel** [comment]
I think this can be merged- already did a review with @sophokles73 earlier today, and this is in all dimensions an improvement to what we had before. More goodness can surely be added, but in subsequent PRs...

File: .github/workflows/check.yaml [hunk 2]
```diff
@@ -0,0 +1,150 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (PLeVasseur) [inline .github/workflows/check.yaml:2]: Update copyright to 2024?

**Review by PLeVasseur — APPROVED** [review]
Impressive stuff! Thanks @AnotherDaniel :slightly_smiling_face: 

I must admit, I am not as knowledgeable as you on the CI, but this looks comprehensive

File: deny.toml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline deny.toml:2]: Since you have added this file newly in 2024 I would suggest to change this accordingly.

File: about.hbs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                    {{/each}}
+                </ul>
+                <pre class="license-text">{{text}}</pre>
+            </li>
+            {{/each}}
+        </ul>
+    </main>
+</body>
+
+</html>
```
Comment (sophokles73) [inline about.hbs:75]: IMHO this line deserves an EOL :-)

File: about.toml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+targets = [
+    "x86_64-unknown-linux-musl",
+    "aarch64-unknown-linux-musl"
+]
+
+ignore-dev-dependencies = true
+
+[unicode-ident]
+accepted = ["Unicode-DFS-2016"]
```
Comment (sophokles73) [inline about.toml:19]: IMHO this line deserves an EOL :-)

File: .github/workflows/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
+# up-rust workflow setup documentation
```
Comment (sophokles73) [inline .github/workflows/README.md:1]: IMHO we do not need this line as the file name alreay suggests that contains documentation and the fact that we renamed the repository multiple times already also makes it questionable if it is a good idea to add the repo name here again ;-)

File: .github/workflows/README.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+## Which automations do we run?
+
+At this time, there are three events that will initiate a workflow run:
+
+### PRs and merges to main
+
+We want a comprehensive but also quick check&test workflow. This should be testing all relevant/obvious feature sets, run on all major OSes, and of course include all the Rust goodness around cargo check, fmt, clippy and so on.
+
+This is implemented in `check.yaml`
```
Comment (sophokles73) [inline .github/workflows/README.md:15]: how about linking to the file?

File: .github/workflows/README.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+### Release publication
+
+We want comprehensive tests and all possible checks, as well as creation of license reports, collection of quality artifacts and publication to crates.io. This workflow pulls in other pieces like the build workflow. An actual release is triggered by pushing a tag that begins with 'v', else this workflow just generates and collects artifacts on workflow level. This will also publish to crates.io if the CRATES_TOKEN secret is set.
+
+This is implemented in `release.yaml`
```
Comment (sophokles73) [inline .github/workflows/README.md:21]: how about linking to the file?

File: .github/workflows/README.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+### Nightly, out of everyone's way
+
+All the tests we can think of, however long they might take. For instance, we can build up-rust for different architectures - this might not really create many insights, but doesn't hurt to try either, and fits nicely into a nightly build schedule.
+
+This is implemented in `nightly.yaml`
```
Comment (sophokles73) [inline .github/workflows/README.md:27]: how about linking to the file?

File: .github/workflows/README.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+## Further workflow modules
+
+In addition to the main workflows described above, there exist a number of modules that are used by these main workflows. They can also be run standalone, and are intendet to make composing the capabilities of our main workflows simpler. These are:
+
+- `coverage.yaml` - collects test code coverage, and can optionally upload the results to codecov.io
```
Comment (sophokles73) [inline .github/workflows/README.md:33]: how about linking to the file?

File: .github/workflows/README.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  - Will publish coverage data to CodeCov if `${{ secrets.CODECOV_TOKEN }}` is set
+  - outputs: download URL for the workflow-generated coverage info file
+- `license-report.yaml` - create a license report for `up-rust` and all its dependencies in html format
```
Comment (sophokles73) [inline .github/workflows/README.md:36]: how about linking to the file?

File: .github/workflows/README.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  - outputs: download URL for the workflow-generated license report
+- `test-featurematrix.yaml` - Test all feature combinations on a range of OS platforms
```
Comment (sophokles73) [inline .github/workflows/README.md:38]: how about linking to the file?

File: .github/workflows/README.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+- `x-build.yaml` - Run release builds on multiple architecture targets
```
Comment (sophokles73) [inline .github/workflows/README.md:39]: how about linking to the file?

File: .github/workflows/check.yaml [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+          cargo nextest run --message-format libtest-json-plus ${{ matrix.feature-flags }} > testresults${{ matrix.feature-flags }}.json
+
+      - name: Upload all-features test results artifact
+        id: test_results
+        if: matrix.feature-flags == '--all-features'
+        uses: actions/upload-artifact@v4
+        with:
+          name: test-results
+          path: testresults--all-features.json
+
```
Comment (sophokles73) [inline .github/workflows/check.yaml:150]: you might want to remove the whitespace from the last line ...

File: .github/workflows/coverage.yaml [hunk 15]
```diff
@@ -0,0 +1,81 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline .github/workflows/coverage.yaml:2]: 2024?

File: .github/workflows/coverage.yaml [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+
+      - name: Upload coverage reports to Codecov
+        env:
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+        if: env.CODECOV_TOKEN != ''
+        uses: codecov/codecov-action@v4.0.1
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          slug: ${{ vars.GITHUB_REPOSITORY }}
+          files: lcov.info
```
Comment (sophokles73) [inline .github/workflows/coverage.yaml:81]: add an EOL?

File: .github/workflows/license-report.yaml [hunk 17]
```diff
@@ -0,0 +1,57 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline .github/workflows/license-report.yaml:2]: 2024?

File: .github/workflows/license-report.yaml [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        cargo install cargo-about
+    - name: Create license report
+      run: |
+        mkdir -p release
+        cargo about generate about.hbs > licenses.html
+    - uses: actions/upload-artifact@v4
+      id: license_report
+      with:
+        name: license-report
+        path: licenses.html
```
Comment (sophokles73) [inline .github/workflows/license-report.yaml:57]: add an EOL?

File: .github/workflows/nightly.yaml [hunk 19]
```diff
@@ -0,0 +1,37 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline .github/workflows/nightly.yaml:2]: 2024?

File: .github/workflows/release.yaml [hunk 20]
```diff
@@ -0,0 +1,177 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline .github/workflows/release.yaml:2]: 2024?

File: .github/workflows/release.yaml [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+    env: 
+      CRATES_TOKEN: ${{ secrets.CRATES_TOKEN }}
+    steps:
+      - uses: actions/checkout@v4
+
+      - if: env.CRATES_TOKEN == ''
+        run: cargo publish --all-features --dry-run
+      - if: env.CRATES_TOKEN != ''
+        run: cargo publish --all-features --token ${CRATES_TOKEN}
+
```
Comment (sophokles73) [inline .github/workflows/release.yaml:177]: you might want to remove the whitespace from the last line

File: .github/workflows/test-featurematrix.yaml [hunk 22]
```diff
@@ -0,0 +1,76 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline .github/workflows/test-featurematrix.yaml:2]: 2024?

File: .github/workflows/x-build.yaml [hunk 23]
```diff
@@ -0,0 +1,67 @@
+# ********************************************************************************
+#  Copyright (c) 2023 Contributors to the Eclipse Foundation
```
Comment (sophokles73) [inline .github/workflows/x-build.yaml:2]: 2024?

**Comment by AnotherDaniel** [comment]
All review comments implemented

**Review by sophokles73 — APPROVED** [review]
LGTM

**Comment by AnotherDaniel** [comment]
Ah, merge it as is - commit history is reasonably clear I'd say.

**Pull request merged**